### PR TITLE
Translate _d_array{,set}ctor to templates

### PR DESCRIPTION
**Pull request by teodutu**
This PR is part of replacing runtime hooks in DMD with templates, as described in [this project](https://github.com/dlang/projects/issues/25).
I also provided a short description of this project in [this forum post](https://forum.dlang.org/post/bbhmgddyabnuxajlxgqp@forum.dlang.org).

This PR concludes Dan Printzell's previous work:
- https://github.com/dlang/dmd/pull/10102
- https://github.com/dlang/druntime/pull/2655

### CODE REVIEW

File: src/dmd/dinterpret.d [hunk 1]
```diff
@@ -4804,6 +4804,41 @@ public:
                 result = interpretRegion(ae, istate);
                 return;
             }
+            else if (fd.ident == Id._d_arrayctor)
```
Comment (thewilsonator) [inline src/dmd/dinterpret.d:4807]: these two branches are almost identical, please merge them

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -9759,6 +9759,100 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                ae.e1.type.nextOf.mutableOf.equals(ae.e2.type.nextOf.mutableOf);
+
+            const isArraySetCtor =
+                (ae.e1.isSliceExp || ae.e1.type.ty == Tsarray) &&
+                ae.e2.isLvalue &&
+                (ae.e2.type.ty == Tstruct || ae.e2.type.ty == Tsarray) &&
+                ae.e1.type.nextOf &&
+                ae.e1.type.nextOf.mutableOf.equals(ae.e2.type.mutableOf);
+
+            if (isArrayCtor)
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d:9793]: these branches also looks very similar, but they are too large to tell at a glance.

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -9759,6 +9759,84 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                ae.e1.type.nextOf &&
+                ae.e2.type.nextOf &&
+                ae.e1.type.nextOf.mutableOf.equals(ae.e2.type.nextOf.mutableOf);
+
+            const isArraySetCtor =
+                (ae.e1.isSliceExp || ae.e1.type.ty == Tsarray) &&
+                ae.e2.isLvalue &&
+                (ae.e2.type.ty == Tstruct || ae.e2.type.ty == Tsarray) &&
+                ae.e1.type.nextOf &&
+                ae.e1.type.nextOf.mutableOf.equals(ae.e2.type.mutableOf);
```
Comment (UplinkCoder) [inline src/dmd/expressionsem.d:9791]: there is an `equivalent`  function in `mtype.d`

**Comment by WalterBright** [comment]
Missing is why do this. What problem is being solved. It adds a bunch of code.

**Comment by PetarKirov** [comment]
@WalterBright You can find more information here: https://forum.dlang.org/thread/bbhmgddyabnuxajlxgqp@forum.dlang.org

@teodutu can you please update the commit message and pull request description with a brief information about the project goals and why the change is necessary?

**Comment by teodutu** [comment]
> @teodutu can you please update the commit message and pull request description with a brief information about the project goals and why the change is necessary?

I mentioned the initial proposal: https://github.com/dlang/projects/issues/25
I hope this provides enough context.

**Comment by RazvanN7** [comment]
@teodutu Rebase this and lets see what happens.

**Comment by RazvanN7** [comment]
```
src/dmd/expressionsem.d' contains trailing whitespace at line 9763
```

**Comment by teodutu** [comment]
This is currently blocked by this issue: https://issues.dlang.org/show_bug.cgi?id=22372

**Comment by RazvanN7** [comment]
@teodutu from my perspective, we need some performance numbers to see how this affects performance. Other than that, I think we're fine.

cc @WalterBright @andralex @atilaneves

**Comment by atilaneves** [comment]
I'm not familiar enough with the existing codebase to be able to comment.

**Comment by teodutu** [comment]
In [this post](https://forum.dlang.org/thread/hajlsppmugslhinluzos@forum.dlang.org), I highlighted that adding a third parameter to `_d_arrayctor` yielded better performance than declaring the returned array inside the hook. For this reason, I chose to go back to the former approach, which is implemented in druntime here: https://github.com/dlang/druntime/pull/3587.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
I'm so happy this finally got in. Congrats for this awesome work @teodutu !

**Comment by andralex** [comment]
Awesome work, thanks! I just took a look at https://github.com/dlang/druntime/pull/2655/files and there are a few opportunities for improvements:

- Instead of doing the memcpy element + postblit loop, we can do just one big memcpy for the whole thing for certain (many!) types. Introspection will figure which is which.
- Types that don't throw from ctors (information we ca n get by introspection) don't need the try/catch. So we're looking at routines tailored for categories of types.
- Could some or all of the work of `enforceRawArraysConformable` be done during compilation?
- `auto element_size = T.sizeof;` is at best redundant and at worse an optimization. Just use `T.sizeof` in the call to `memcpy` to make sure the compiler sees it's a compile-time constant.

**Comment by radcapricorn** [comment]
To add to Andrei's notes, if these functions are going to be instantiated by user code, the try/catches also need a `version(D_BetterC)` no-op.

**Pull request merged**